Rotate log files and prune oldest archival files

pinheadmz opened the pull request:
API changes:

`Logger` now parses two extra options:
- `maxFileSize` (default `20 << 20` or ~20 MB): The maximum file size allowed before starting a new file.
- `maxFiles` (default `10`): The number of archival files to keep on disk (older files are pruned)

Behavior changes:

`Logger` now keeps track of how much data is being written to the current log file. When that data reaches the `maxFileSize`, the new method `rotate()` is called. `rotate()` CLOSES the logger, renames the current log file, and then REOPENS the logger with a fresh `debug.log` file of 0 bytes. The archival file is renamed with the current timestamp, e.g. `debug_2019-10-30_14-49-23.log`. Once the rotation is complete, a new method `prune()` is called. `prune()` gets the names of all the files in the `prefix` directory that match the above naming scheme by regex. The archival files are sorted alphabetically which (due to the timestamps) will be sorted oldest to newest. The newest files up to `maxFiles` are kept and the rest are deleted from disk.

Result:

I ran `hsd` with a modified version of this branch, restricting `maxFileSize` to `1 << 20` and `maxFiles` to `4`. After finishing sync on testnet my data directory looked like this:

```
$ ls -lah
total 9648
drwxr-xr-x  12 matthewzipkin  staff   384B Oct 30 10:50 .
drwxr-xr-x  11 matthewzipkin  staff   352B Oct 30 10:48 ..
drwxr-xr-x  47 matthewzipkin  staff   1.5K Oct 30 10:49 chain
-rw-r--r--   1 matthewzipkin  staff   613K Oct 30 10:50 debug.log
-rw-r--r--   1 matthewzipkin  staff   1.0M Oct 30 10:49 debug_2019-10-30_14-49-18.log
-rw-r--r--   1 matthewzipkin  staff   1.0M Oct 30 10:49 debug_2019-10-30_14-49-23.log
-rw-r--r--   1 matthewzipkin  staff   1.0M Oct 30 10:49 debug_2019-10-30_14-49-34.log
-rw-r--r--   1 matthewzipkin  staff   1.0M Oct 30 10:49 debug_2019-10-30_14-49-49.log
-rw-r--r--   1 matthewzipkin  staff    30K Oct 30 10:50 hosts.json
-rw-r--r--   1 matthewzipkin  staff    65B Oct 30 10:48 key
drwxr-x---   5 matthewzipkin  staff   160B Oct 30 10:48 tree
drwxr-xr-x   7 matthewzipkin  staff   224B Oct 30 10:48 wallet
```

Known issue:

When `logger` is rotating, messages can not be written to disk and are just lost. During IBD, messages are actually being written so fast that with extreme settings like my example above, several lines are completely lost in between files. (I estimate around 10 lines were lost based on what was kept at the end of one file and the beginning of the next). I don't think this issue will be too devastating in practice however, given the default `maxFileSize` and how rarely a rotation will occur. Users can use the new options to prevent rotation altogether. Using the included test, I was able to determine that on my system at least, a 5ms pause between log lines is all it took to not lose any data between files.

pinheadmz committed:
logger: get and keep track of file size

pinheadmz committed:
logger: rotate out current log file

pinheadmz committed:
test: refactor for better test separation

pinheadmz committed:
logger: rotate log file at size limit

pinheadmz committed:
logger: prune oldest archival files

pinheadmz commented:
As an afterthought -- we can now also remove `truncate()` from `logger`, or at least the call to `truncate()` during `open()`. That method explicitly deletes log data from disk that may be relevant to the user. Thoughts on that?

braydonf commented:
> When logger is rotating, messages can not be written to disk and are just lost. 

This is a problem that should be addressed.

tynes commented:
Maybe its possible to buffer the log lines in memory while the logger is rotating.

pinheadmz commented:
@tynes attempting that here: https://github.com/bcoin-org/blgr/pull/8/commits/127d353dffee05c4762bdc8c7f9842c7ffa3c6a3

Still losing lots of data when rotating.

pinheadmz committed:
logger: timestamp archival files with ms

This is mainly needed in testing when log files are generated
so fast that the filenames collide and files get replaced.

pinheadmz committed:
test: add time in milliseconds to logged messages

pinheadmz committed:
logger: buffer output while rotating

pinheadmz commented:
Updates since last review:

- Archival file names now include ms in the timestamp:
`debug_2019-11-12_20-56-30-408.log` (year-month-day_hour-min-sec-ms)
This was required for the test which was printing files so quickly they were getting over written because they had the same timestamp in seconds.

- During file rotation, output is written to a buffer so no data is lost.

I played with the test case which writes lines to the log in a for loop. Before these updates, tests would fail without a 5ms timeout between log lines. With the buffer in place, this timeout can be reduced to 0 and no data is lost. However, with no `setTimeout` at all, or inserting a `setImmediate` instead, data *is* still lost and the test will fail. I consider this a big improvement! But I'm not sure how tight we want this to be.

Putting into practice, I reran my Handshake testnet test with the same parameters, and observed NO LOSS in data, every line was printed to consecutive files.

pinheadmz committed:
logger: remove shrink option and truncate function

These methods are deprecated now that we can rotate log files
instead of truncating at startup and deleting important data.

pinheadmz commented:
Changelog updated and package version bumped -- although I just bumped the patch version. Should this be a more significant version bump due to breaking API change (removing `shrink` and `truncate()`?)

braydonf commented:
With [semantic versioning](https://semver.org/) a breaking change is a _major_ version bump, an addition to the api is a _minor_ and a fix is a _patch_ version.

braydonf commented:
Also, it's probably not worth it to bump the version in the PR, that can be done at release and tag time.

pinheadmz commented:
Yeah I dunno why I tried to include that. I just lopped off the version bump commit this PR

pinheadmz committed:
README: Add API change
